### PR TITLE
[mds-stream] [mds-event-processor] NATS implementation for mds-event-processor.

### PR DESCRIPTION
**Pull request by avatarneil**
Implementation of NATS for streaming events to `mds-event-processor`. This PR removes KNE from our stack, as well as delegating to regular Kubernetes CronJobs for batch processing. Has been tested locally and on GKE and EKS. 

## PR Checklist

 - [x] simple searchable title - `[mds-db] Add PG env var`, `[config] Fix eslint config`
 - [x] briefly describe the changes in this PR
 - [ ] write tests for all new functionality

## Impacts
- [ ] Provider
- [x] Agency
- [ ] Audit
- [ ] Policy
- [ ] Compliance
- [ ] Daily
- [ ] Native
- [ ] Policy Author

### CODE REVIEW

**Comment by wesleyfsmith** [comment]
Can I get some more context on this PR? I'm not sure I'm following

**Comment by avatarneil** [comment]
@wesleyfsmith Sure, basically at the moment we're using both NATS and Knative Eventing to send data from `mds-agency -> mds-event-processor`. What we've found trying to get the current solution running in a remote cluster is a whole bundle of ops headaches, so we're trying out a few different alternatives to KNE to see if they work better (at least in the short-term). I have a handful of other personal withholdings against KNE, such as the fact that it's in `v0.11` right now...

**Comment by jwtodd** [comment]
let's start out by locally deploying a nats (vs streaming nats, aka stan) process:

```
git clone https://github.com/nats-io/k8s
cd k8s
kubectl config use-context docker-desktop
./setup.sh --without-tls --without-auth
```

we can then `top` the nats process by using the nats-clis found in the nats-box pod:

```
kubectl exec -it nats-box -- /bin/sh
nats-top -s nats
```

next we can test pub/sub to the nats server, again, via nats-box:

```
kubectl exec -it nats-box -- /bin/sh
nats-pub test &
nats-sub test 'hello world'
```

now with this in place we should be able to connect nats client code to the host `nats.default` ... note: we should verify the fully-qualified nats host name that spans namespaces.

**Comment by jwtodd** [comment]
i believe we can also do the same connection against the natss process we deploy today with mdsctl ... but the above process is closer to the ultimate target process.

**Comment by jwtodd** [comment]
to start out, we could port-forward nats to test the local client, eg:

```
kubectl port-forward nats-0 4222 &
```

**Comment by jwtodd** [comment]
@avatarneil added nats, disabled kne/natss, config'd env NATS via helm/mds for service lookup-by-name

bootstrap will install: helm, tiller, istio, nats

note: might need to uninstall kne/natss by hand (or k8s factory reset); eg: ./bin/mdsctl uninstall:knative,natss

**Comment by janedotx** [comment]
@avatarneil So how are you liking this? The actual TS code looks pretty straightforward, like the kubemq code.

**Comment by jwtodd** [comment]
additionally, the top-level README.md has been updated with more kubernetes how-to notes.

one lil patch i am working on atm is to remove the mdsctl persistence to ~/.bashrc ... but this is not a blocker per se.

let's do this :)

File: helm/mds/templates/configmap.yaml [hunk 1]
```diff
@@ -5,3 +5,21 @@ metadata:
...
+    -----BEGIN NATS USER JWT-----
+    eyJ0eXAiOiJqd3QiLCJhbGciOiJlZDI1NTE5In0.eyJqdGkiOiJDT0JKTUU3Skc0M1dLNVNKNEhPVjdUM01TUUlQUUVZU0lVTUJYTjVIUTYzVE1IQTdERlJRIiwiaWF0IjoxNTc4Njg1ODYwLCJpc3MiOiJBQlBQQkk3V0hQS0tVQUQ0RE5NQ1c2V09STjdNSVNPSE1IUTVZSU9ISk5RM0M1VFRPTzRDV0ZaTyIsIm5hbWUiOiJzeXMiLCJzdWIiOiJVQTRUSkU1M1hFVEtGUFRFQ1lWRVpBSzNIMkVDR1VSSVdZSkpaRVREUU81QVhSNkhWN0ZENUhJViIsInR5cGUiOiJ1c2VyIiwibmF0cyI6eyJwdWIiOnt9LCJzdWIiOnt9fX0.2jwAmP62UcMh0l3fL4Z-cJAXl1guG61qqb0IhXVNZBqPZyxvfgvYZderN10c-Be0wC-m9P-2gY6HWngiHU3sAw
+    ------END NATS USER JWT------
+    ************************* IMPORTANT *************************
+    NKEY Seed printed below can be used to sign and prove identity.
+    NKEYs are sensitive and should be treated as secrets.
+    -----BEGIN USER NKEY SEED-----
+    SUAGIRH2Z24KWAG5BFFUJ732P7KRFIYZBGERJXVGOZCFNVSNNCVAZDZU2A
+    ------END USER NKEY SEED------
+    *************************************************************
```
Comment (invertigo) [inline helm/mds/templates/configmap.yaml]: @jwtodd - do we still need this?

File: helm/mds/templates/cronjobsource.yaml [hunk 2]
```diff
@@ -1,17 +1,17 @@
...
+# spec:
+#   schedule: {{ $processor.triggeredBy.cron | quote }}
+#   sink:
+#     ref:
+#       apiVersion: v1
+#       kind: Service
+#       name: {{ $name }}
+# ---
+# {{- end }}
+# {{- end }}
```
Comment (invertigo) [inline helm/mds/templates/cronjobsource.yaml]: @jwtodd this is for knative eventing implementation, no?  Can it be removed?

File: helm/mds/templates/deployment.yaml [hunk 3]
```diff
@@ -22,12 +22,15 @@ spec:
         - name: mds-config-files
           configMap:
             name: mds-config-files
+        - name: stan-creds-volume
+          configMap:
+            name: stans-creds
+          # secret:
+          #   secretName: stan-creds
```
Comment (invertigo) [inline helm/mds/templates/deployment.yaml]: @jwtodd no longer used?

File: helm/mds/templates/deployment.yaml [hunk 4]
```diff
@@ -44,6 +47,8 @@ spec:
         volumeMounts:
         - name: mds-config-files
           mountPath: /mds-config/
+        - name: stan-creds-volume
+          mountPath: /var/run/nats/creds/stan
```
Comment (invertigo) [inline helm/mds/templates/deployment.yaml]: same

File: helm/mds/templates/deployment.yaml [hunk 5]
```diff
@@ -86,15 +91,17 @@ spec:
...
             value: {{ $.Values.redis.port | quote }}
-          - name: EVENT_SOURCE
-            value: cloudevents
-          - name: CE_NAME
-            value: {{ $name }}
-          - name: SINK
-            value: http://natss-broker.{{ $.Release.Namespace }}.svc.cluster.local
-          {{- if not (empty $.Values.tenantId) }}
+          # - name: EVENT_SOURCE
+          #   value: cloudevents
```
Comment (invertigo) [inline helm/mds/templates/deployment.yaml]: @jwtodd cleanup

File: helm/mds/templates/natss-broker.yaml [hunk 6]
```diff
@@ -1,11 +1,11 @@
...
+# kind: Broker
+# metadata:
+#   name: natss # better name
+#   namespace: {{ $.Release.Namespace }}
+# spec:
+#   channelTemplateSpec:
+#     apiVersion: messaging.knative.dev/v1alpha1
+#     kind: NatssChannel
+# #    kind: ClusterChannelProvisioner
+# #    name: natss
```
Comment (invertigo) [inline helm/mds/templates/natss-broker.yaml]: @jwtodd cleanup

File: helm/mds/templates/natss-channel.yaml [hunk 7]
```diff
@@ -1,5 +1,5 @@
-apiVersion: messaging.knative.dev/v1alpha1
-kind: NatssChannel
-metadata:
-  name: natss
-  namespace: {{ $.Release.Namespace }}
+# apiVersion: messaging.knative.dev/v1alpha1
+# kind: NatssChannel
+# metadata:
+#   name: natss
+#   namespace: {{ $.Release.Namespace }}
```
Comment (invertigo) [inline helm/mds/templates/natss-channel.yaml]: @jwtodd cleanup

File: helm/mds/templates/processor-deployment.yaml [hunk 8]
```diff
@@ -22,6 +22,11 @@ spec:
         - name: mds-config-files
           configMap:
             name: mds-config-files
+        - name: stan-creds-volume
+          configMap:
+            name: stans-creds
+          # secret:
+          #   secretName: stan-creds
```
Comment (invertigo) [inline helm/mds/templates/processor-deployment.yaml]: @jwtodd cleanup

File: helm/mds/templates/processor-deployment.yaml [hunk 9]
```diff
@@ -86,12 +93,12 @@ spec:
...
           - name: REDIS_PORT
             value: {{ $.Values.redis.port | quote }}
-          - name: EVENT_SOURCE
-            value: cloudevents
-          - name: CE_NAME
-            value: {{ $name }}
-          - name: SINK
-            value: http://natss-broker.{{ $.Release.Namespace }}.svc.cluster.local
+          # - name: EVENT_SOURCE
+          #   value: cloudevents
```
Comment (invertigo) [inline helm/mds/templates/processor-deployment.yaml]: cleanup

File: helm/mds/values.yaml [hunk 10]
```diff
@@ -57,17 +57,17 @@ processors:
...
-      cron: '0 * * * *'
+      cron: '* * * * *'
   mds-provider-processor:
-    enabled: true
+    enabled: false
     version: latest
     migration: false
     triggeredBy:
-      cron: '0 * * * *'
+      cron: '* * * * *'
```
Comment (invertigo) [inline helm/mds/values.yaml]: @jwtodd I believe we can completely remove the mds-[trip/event]-processor Deployments now.

File: helm/prometheus-operator/Chart.yaml [hunk 11]
```diff
@@ -0,0 +1,5 @@
+apiVersion: v1
+appVersion: '0.4.0'
+description: Prometheus Operator
+name: prometheus-operator
+version: 0.0.1
```
Comment (invertigo) [inline helm/prometheus-operator/Chart.yaml]: @jwtodd Is this related to nats at all?  should it be in this PR?

File: helm/prometheus-operator/templates/prometheus-operator.yml [hunk 12]
```diff
@@ -0,0 +1,207 @@
+---
```
Comment (invertigo) [inline helm/prometheus-operator/templates/prometheus-operator.yml]: @jwtodd likewise this - should it be in this PR?

File: helm/mds/templates/cronjobsource.yaml (hunk 2 above)
Comment (jwtodd) [reply helm/mds/templates/cronjobsource.yaml]: will do

File: helm/mds/values.yaml (hunk 10 above)
Comment (jwtodd) [reply helm/mds/values.yaml]: yeah but we still use this section of values.yaml to configure the k8s CronJob pods (see: helm/mds/processors-cronjob.yaml)

a measured amount of cleanup should be done here. @evanxadkins thoughts?

my $.02US ... have a simple templates/cronjob.yaml for all cronjob managed things

File: helm/prometheus-operator/Chart.yaml (hunk 11 above)
Comment (jwtodd) [reply helm/prometheus-operator/Chart.yaml]: yeah ... i'd like to get grafana/surveyor back on the table. testing/verifying locally atm

File: helm/mds/values.yaml (hunk 10 above)
Comment (jwtodd) [reply helm/mds/values.yaml]: @evanxadkins also, do we want to set the interval back to every hour?

**Comment by jwtodd** [comment]
@invertigo removed all the mds/stan dead code with the exception of helm/mds/values.yaml ... which i think could use some general cleanup in how best we should manage cronjob resources. @evanxadkins want to talk this through a bit? i can haz ideas.

File: helm/prometheus-operator/Chart.yaml (hunk 11 above)
Comment (invertigo) [reply helm/prometheus-operator/Chart.yaml]: Maybe a new PR for that then?  Seems like this isn't a blocker for NATS and the event layer.

Comment (invertigo) [reply helm/prometheus-operator/Chart.yaml]: or is it specifically for monitoring NATS message flow?

File: container-images/mds-provider-processor/index.ts [hunk 13]
```diff
@@ -14,14 +14,15 @@
...
 
-const { npm_package_name, npm_package_version, npm_package_git_commit, PORT = 5001 } = env()
-
-EventServer(processor).listen(PORT, () =>
-  /* eslint-reason avoids import of logger */
-  /* eslint-disable-next-line no-console */
-  console.log(`${npm_package_name} v${npm_package_version} (${npm_package_git_commit}) running on port ${PORT}`)
-)
+// eslint-disable-next-line @typescript-eslint/no-floating-promises
+providerProcessor()
```
Comment (drtyh2o) [inline container-images/mds-provider-processor/index.ts]: Next TypeScript version is supposed to support top-level `await` .... :-)

File: helm/mds/values.yaml [hunk 14]
```diff
@@ -3,7 +3,7 @@ apis:
     enabled: true
     pathPrefix: /agency
     version: latest
-    migration: false
+    migration: true
```
Comment (drtyh2o) [inline helm/mds/values.yaml]: Probably don't want to commit this....

File: helm/mds/values.yaml (hunk 10 above)
Comment (drtyh2o) [reply helm/mds/values.yaml]: This should not be `* * * * *` (every minute) I think.

File: packages/mds-event-server/index.ts [hunk 15]
```diff
@@ -1,18 +1,123 @@
...
+  NATS,
+  STAN_CLUSTER,
+  STAN_CREDS
+}: {
+  NATS: string
+  STAN_CLUSTER: string
+  STAN_CREDS?: string
+}) => {
+  return stan.connect(STAN_CLUSTER, `mds-event-processor-${uuid()}`, {
+    url: `nats://${NATS}:4222`,
```
Comment (drtyh2o) [inline packages/mds-event-server/index.ts]: Should this `url` also be constructed from an environment variable? ... using `NATS_PORT` perhaps?

File: container-images/mds-provider-processor/index.ts (hunk 13 above)
Comment (avatarneil) [reply container-images/mds-provider-processor/index.ts]: So excited 😍

File: packages/mds-event-server/index.ts (hunk 15 above)
Comment (jwtodd) [reply packages/mds-event-server/index.ts]: +1 let's put the entire url in helm ... more declaratively obvious, where other like resources reside, etc.

File: helm/mds/values.yaml (hunk 14 above)
Comment (invertigo) [reply helm/mds/values.yaml]: Fixed

File: container-images/mds-event-processor/index.ts [hunk 16]
```diff
@@ -14,14 +15,16 @@
...
+} = process
 
-EventServer(processor).listen(PORT, () =>
-  /* eslint-reason avoids import of logger */
-  /* eslint-disable-next-line no-console */
-  console.log(`${npm_package_name} v${npm_package_version} (${npm_package_git_commit}) running on port ${PORT}`)
+Promise.all([initializeStanSubscriber({ NATS, STAN_CLUSTER, STAN_CREDS, TENANT_ID, processor }), EventServer()]).catch(
+  // eslint-disable-next-line promise/prefer-await-to-callbacks
+  err => {
+    console.log(err)
```
Comment (janedotx) [inline container-images/mds-event-processor/index.ts]: Why a `console.log` and not the `mds-logger`?

File: packages/mds-stream/index.ts [hunk 17]
```diff
@@ -168,23 +193,23 @@ async function writeStreamBatch(stream: Stream, field: string, values: unknown[]
 
 // put basics of vehicle in the cache
 async function writeDevice(device: Device) {
-  if (env.SINK) {
-    return writeCloudEvent('device', JSON.stringify(device))
+  if (env.NATS) {
+    return writeNatsEvent('device', JSON.stringify(device))
```
Comment (janedotx) [inline packages/mds-stream/index.ts]: Should `writeCloudEvent` be deleted now? I don't see any calls to it.

File: container-images/mds-event-processor/index.ts (hunk 16 above)
Comment (avatarneil) [reply container-images/mds-event-processor/index.ts]: This avoids the import of `mds-logger` at the container level, and is a pattern we use with all of our other apps.

File: packages/mds-stream/index.ts (hunk 17 above)
Comment (avatarneil) [reply packages/mds-stream/index.ts]: Added a comment explaining why we're keeping it still
